File: src/libs/m3/vfs/GenericFile.cc
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Log.h>

#include <m3/Syscalls.h>
#include <m3/com/GateStream.h>
#include <m3/com/OpCodes.h>
#include <m3/session/M3FS.h>
#include <m3/vfs/FileTable.h>
#include <m3/vfs/GenericFile.h>
#include <m3/vfs/VFS.h>

namespace m3 {

GenericFile::GenericFile(int flags, capsel_t caps, size_t fs_id, size_t id, epid_t mep,
                         SendGate *sg)
    : File(flags),
      _fs_id(fs_id),
      _id(id),
      _sess(caps + 0, sg ? ObjCap::KEEP_CAP : 0),
      _sg(sg ? sg : new SendGate(SendGate::bind(caps + 1))),
      _notify_rgate(),
      _notify_sgate(),
      _notify_received(),
      _notify_requested(),
      _mg(MemGate::bind(ObjCap::INVALID)),
      _goff(),
      _off(),
      _pos(),
      _len(),
      _writing() {
    if(mep != TCU::INVALID_EP)
        _mg.set_ep(new EP(EP::bind(mep)));
}

GenericFile::~GenericFile() {
    if(have_sess())
        delete _sg;
    else {
        // we never want to invalidate the EP
        delete const_cast<EP *>(_mg.ep());
        _mg.set_ep(nullptr);
    }
}

void GenericFile::remove() noexcept {
    LOG(LogFlags::LibFS, "GenFile[{}]::evict()"_cf, _fd);

    // commit read/written data
    try {
        if(_writing)
            commit();
    }
    catch(...) {
        // ignore
    }

    if(!have_sess()) {
        auto fs = Activity::own().mounts()->get_by_id(_fs_id);
        if(fs)
            fs->close(_id);
    }
    else {
        try {
            const EP *ep = _mg.ep();
            if(ep)
                Activity::own().revoke(KIF::CapRngDesc(KIF::CapRngDesc::OBJ, ep->sel()), true);
        }
        catch(...) {
            // ignore
        }

        try {
            Activity::own().revoke(KIF::CapRngDesc(KIF::CapRngDesc::OBJ, _sess.sel()));
        }
        catch(...) {
            // ignore
        }
    }
}

Errors::Code GenericFile::try_stat(FileInfo &info) const {
    LOG(LogFlags::LibFS, "GenFile[{}]::stat()"_cf, _fd);

    GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::FSTAT, _id);
    Errors::Code res;
    reply >> res;
    if(res == Errors::SUCCESS)
        reply >> info;
    return res;
}

size_t GenericFile::seek(size_t offset, int whence) {
    LOG(LogFlags::LibFS, "GenFile[{}]::seek({}, {})"_cf, _fd, offset, whence);

    // handle SEEK_CUR as SEEK_SET
    if(whence == M3FS_SEEK_CUR) {
        offset = _goff + _off + _pos + offset;
        whence = M3FS_SEEK_SET;
    }

    // try to seek locally first
    if(whence == M3FS_SEEK_SET) {
        // no change?
        if(offset == _goff + _off + _pos)
            return offset;

        // first commit the written data
        if(_writing)
            commit();

        if(offset >= _goff + _off && offset <= _goff + _off + _len) {
            _pos = offset - (_goff + _off);
            return offset;
        }
    }
    else {
        // first commit the written data
        if(_writing)
            commit();
    }

    // now seek on the server side
    GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::SEEK, _id, offset, whence);
    reply.pull_result();

    reply >> _goff >> _off;
    _pos = _len = 0;
    return _goff + _off;
}

std::string GenericFile::path() {
    std::string path;
    GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::GET_PATH, _id);
    reply.pull_result();
    reply >> path;

    const char *mount = Activity::own().mounts()->path_of_id(_fs_id);

    OStringStream abspath;
    format_to(abspath, "{}/{}"_cf, mount, path);
    return abspath.str();
}

void GenericFile::truncate(size_t length) {
    if(_writing)
        commit();

    GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::TRUNCATE, _id, length);
    reply.pull_result();
    // reset position in case we were behind the truncated position
    reply >> _goff >> _off;
    // we've lost access to the previous extent
    _pos = _len = 0;
}

NOINLINE bool GenericFile::receive_notify(uint event, bool fetch) {
    // not received the event yet?
    if((_notify_received & event) == 0) {
        // if we did not request a notification for this event yet, do that now
        if((_notify_requested & event) == 0)
            request_notification(event);

        const TCU::Message *msg;
        // if there is a message, add it to the received events
        if((msg = _notify_rgate->fetch()) != nullptr) {
            uint events;
            GateIStream imsg(*_notify_rgate, msg);
            imsg >> events;
            _notify_received |= events;
            _notify_requested &= ~events;
            LOG(LogFlags::LibFS, "GenFile[{}]::receive_notify() -> received {:x}"_cf, _fd, events);
            // give credits back to sender
            reply_vmsg(imsg, 0);
        }
    }

    // now check again if we have received this event; if not, we would block
    if((_notify_received & event) == 0)
        return false;

    if(fetch) {
        // okay, event received; remove it and continue
        LOG(LogFlags::LibFS, "GenFile[{}]::receive_notify() -> fetched {:x}"_cf, _fd, event);
        _notify_received &= ~event;
    }

    return true;
}

Option<size_t> GenericFile::read(void *buffer, size_t count) {
    delegate_ep();
    if(_writing)
        commit();

    LOG(LogFlags::LibFS, "GenFile[{}]::read({}, pos={})"_cf, _fd, count, _goff + _pos);

    if(_pos == _len) {
        if(!_blocking && !receive_notify(Event::INPUT, true))
            return None;

        GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::NEXT_IN, _id);
        Errors::Code res;
        reply >> res;
        // if the server promised that we can call NEXT_IN without being blocked, but would still
        // have to block us, it returns Errors::WOULD_BLOCK instead.
        if(res == Errors::WOULD_BLOCK)
            return None;
        if(res != Errors::SUCCESS)
            throw Exception(res);

        _goff += _len;
        reply >> _off >> _len;
        _pos = 0;
    }

    size_t amount = Math::min(count, _len - _pos);
    if(amount > 0) {
        if(flags() & FILE_NODATA) {
            if(count > 2)
                CPU::compute(count / 2);
        }
        else
            _mg.read(buffer, amount, _off + _pos);
        _pos += amount;
    }
    return Some(amount);
}

Option<size_t> GenericFile::write(const void *buffer, size_t count) {
    delegate_ep();

    LOG(LogFlags::LibFS, "GenFile[{}]::write({}, pos={})"_cf, _fd, count, _goff + _pos);

    if(_pos == _len) {
        if(!_blocking && !receive_notify(Event::OUTPUT, true))
            return None;

        GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::NEXT_OUT, _id);
        Errors::Code res;
        reply >> res;
        // if the server promised that we can call NEXT_OUT without being blocked, but would still
        // have to block us, it returns Errors::WOULD_BLOCK instead.
        if(res == Errors::WOULD_BLOCK)
            return None;
        if(res != Errors::SUCCESS)
            throw Exception(res);

        _goff += _len;
        reply >> _off >> _len;
        _pos = 0;
    }

    size_t amount = Math::min(count, _len - _pos);
    if(amount > 0) {
        if(flags() & FILE_NODATA) {
            if(count > 4)
                CPU::compute(count / 4);
        }
        else
            _mg.write(buffer, amount, _off + _pos);
        _pos += amount;
    }
    _writing = true;
    return Some(amount);
}

void GenericFile::commit() {
    if(_pos > 0) {
        LOG(LogFlags::LibFS, "GenFile[{}]::commit({}, {})"_cf, _fd, _writing ? "write" : "read",
            _pos);

        GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::COMMIT, _id, _pos);
        reply.pull_result();

        // if we append, the file was truncated
        _goff += _pos;
        _pos = _len = 0;
    }
    _writing = false;
}

void GenericFile::sync() {
    commit();

    LOG(LogFlags::LibFS, "GenFile[{}]::sync()"_cf, _fd);
    GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::SYNC, _id);
    reply.pull_result();
}

Errors::Code GenericFile::try_get_tmode(File::TMode *mode) noexcept {
    GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::GET_TMODE, _id);
    Errors::Code res;
    reply >> res;
    if(res != Errors::SUCCESS)
        return res;
    reply >> *mode;
    return Errors::SUCCESS;
}

void GenericFile::set_tmode(TMode mode) {
    GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::SET_TMODE, _id, mode);
    reply.pull_result();
}

NOINLINE void GenericFile::enable_notifications() {
    if(_notify_rgate)
        return;

    std::unique_ptr<RecvGate> notify_rgate(new RecvGate(
        RecvGate::create(nextlog2<NOTIFY_MSG_SIZE>::val, nextlog2<NOTIFY_MSG_SIZE>::val)));
    notify_rgate->activate();

    std::unique_ptr<SendGate> notify_sgate(new SendGate(SendGate::create(&*notify_rgate)));

    KIF::ExchangeArgs args;
    ExchangeOStream os(args);
    os << opcodes::File::ENABLE_NOTIFY;
    args.bytes = os.total();
    KIF::CapRngDesc crd(KIF::CapRngDesc::OBJ, notify_sgate->sel(), 1);
    _sess.delegate_for(Activity::own(), crd, &args);

    LOG(LogFlags::LibFS, "GenFile[{}]::enable_notifications()"_cf, _fd);

    // now that it succeeded, store the gates
    _notify_rgate.swap(notify_rgate);
    _notify_sgate.swap(notify_sgate);
}

void GenericFile::request_notification(uint events) {
    LOG(LogFlags::LibFS, "GenFile[{}]::request_notification(want={:x}, have={:x})"_cf, _fd, events,
        _notify_requested);

    if((_notify_requested & events) != events) {
        GateIStream reply = send_receive_vmsg(*_sg, opcodes::File::REQ_NOTIFY, _id, events);
        reply.pull_result();
        _notify_requested |= events;
    }
}

bool GenericFile::check_events(uint events) {
    if(_blocking)
        return true;
    return receive_notify(events, false);
}

bool GenericFile::fetch_signal() {
    if(!_notify_rgate)
        enable_notifications();

    return receive_notify(Event::SIGNAL, true);
}

void GenericFile::map(Reference<Pager> &pager, goff_t *virt, size_t fileoff, size_t len, int prot,
                      int flags) const {
    pager->map_ds(virt, len, prot, flags, _sess, fileoff);
}

FileRef<File> GenericFile::clone() const {
    if(!have_sess())
        throw Exception(Errors::NOT_SUP);

    KIF::CapRngDesc crd(KIF::CapRngDesc::OBJ, Activity::own().alloc_sels(2), 2);
    do_clone(Activity::own(), crd);
    auto file = std::unique_ptr<File>(new GenericFile(flags(), crd.start(), _fs_id));
    return Activity::own().files()->alloc(std::move(file));
}

void GenericFile::do_clone(Activity &act, KIF::CapRngDesc &crd) const {
    KIF::ExchangeArgs args;
    ExchangeOStream os(args);
    os << opcodes::File::CLONE_FILE;
    args.bytes = os.total();
    _sess.obtain_for(act, crd, &args);
}

void GenericFile::delegate_ep() {
    if(!_mg.ep()) {
        const EP &ep = _mg.acquire_ep();
        do_delegate_ep(ep);
    }
}

void GenericFile::do_delegate_ep(const EP &ep) const {
    LOG(LogFlags::LibFS, "GenFile[{}]::delegate_ep({})"_cf, _fd, ep.id());

    KIF::ExchangeArgs args;
    ExchangeOStream os(args);
    os << opcodes::File::SET_DEST;
    args.bytes = os.total();
    KIF::CapRngDesc crd(KIF::CapRngDesc::OBJ, ep.sel(), 1);
    _sess.delegate_for(Activity::own(), crd, &args);
}

}

File: src/libs/rust/m3/src/vfs/genericfile.rs
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use core::any::Any;
use core::cmp;
use core::fmt;

use crate::boxed::Box;
use crate::cap::Selector;
use crate::col::{String, ToString};
use crate::com::GateIStream;
use crate::com::{recv_reply, MemGate, RecvGate, SendGate, EP};
use crate::errors::{Code, Error};
use crate::goff;
use crate::int_enum;
use crate::io::{Read, Write};
use crate::kif::{CapRngDesc, CapType, Perm, INVALID_SEL};
use crate::math;
use crate::rc::Rc;
use crate::serialize::Source;
use crate::session::{ClientSession, HashInput, HashOutput, HashSession, MapFlags, Pager};
use crate::tcu::EpId;
use crate::tiles::{Activity, ChildActivity, StateSerializer};
use crate::vfs::{
    filetable, Fd, File, FileEvent, FileInfo, Map, OpenFlags, Seek, SeekMode, StatResponse,
};

int_enum! {
    /// The operations for [`GenericFile`].
    pub struct GenFileOp : u64 {
        const STAT          = 0;
        const SEEK          = 1;
        const NEXT_IN       = 2;
        const NEXT_OUT      = 3;
        const COMMIT        = 4;
        const TRUNCATE      = 5;
        const SYNC          = 6;
        const CLOSE         = 7;
        const CLONE         = 8;
        const GET_PATH      = 9;
        const SET_TMODE     = 10;
        const SET_DEST      = 11;
        const ENABLE_NOTIFY = 12;
        const REQ_NOTIFY    = 13;
    }
}

const NOTIFY_MSG_SIZE: usize = 64;

struct NonBlocking {
    notify_rgate: Box<RecvGate>,
    _notify_sgate: Box<SendGate>,
    notify_received: FileEvent,
    notify_requested: FileEvent,
}

/// A file implementation for all file-like objects.
///
/// `GenericFile` implements the file protocol and can therefore be used for m3fs files, pipes,
/// virtual terminals, and whatever else provides file-like objects in the future.
pub struct GenericFile {
    id: Option<usize>,
    fs_id: Option<usize>,
    fd: Fd,
    flags: OpenFlags,
    sess: ClientSession,
    sgate: Rc<SendGate>,
    mgate: MemGate,
    delegated_ep: Selector,
    blocking: bool,
    nb_state: Option<NonBlocking>,
    goff: usize,
    off: usize,
    pos: usize,
    len: usize,
    writing: bool,
}

impl GenericFile {
    pub(crate) fn new(flags: OpenFlags, sel: Selector, fs_id: Option<usize>) -> Self {
        GenericFile {
            id: None,
            fs_id,
            fd: filetable::INV_FD,
            flags,
            sess: ClientSession::new_bind(sel),
            sgate: Rc::new(SendGate::new_bind(sel + 1)),
            mgate: MemGate::new_bind(INVALID_SEL),
            delegated_ep: INVALID_SEL,
            blocking: true,
            nb_state: None,
            goff: 0,
            off: 0,
            pos: 0,
            len: 0,
            writing: false,
        }
    }

    pub(crate) fn new_without_sess(
        flags: OpenFlags,
        sel: Selector,
        id: usize,
        fs_id: usize,
        mep: EpId,
        sgate: Rc<SendGate>,
    ) -> Self {
        let mut mgate = MemGate::new_bind(INVALID_SEL);
        mgate.set_ep(Some(EP::new_bind(mep, INVALID_SEL)));
        GenericFile {
            id: Some(id),
            fs_id: Some(fs_id),
            fd: filetable::INV_FD,
            flags,
            sess: ClientSession::new_bind(sel),
            sgate,
            mgate,
            delegated_ep: INVALID_SEL,
            blocking: true,
            nb_state: None,
            goff: 0,
            off: 0,
            pos: 0,
            len: 0,
            writing: false,
        }
    }

    fn file_id(&self) -> usize {
        self.id.unwrap_or(0)
    }

    pub(crate) fn unserialize(s: &mut Source<'_>) -> Box<dyn File> {
        let flags: u32 = s.pop().unwrap();
        let sel: Selector = s.pop().unwrap();
        let fs_id: usize = s.pop().unwrap();
        Box::new(GenericFile::new(
            OpenFlags::from_bits_truncate(flags),
            sel,
            if fs_id == !0 { None } else { Some(fs_id) },
        ))
    }

    fn submit(&mut self, force: bool) -> Result<(), Error> {
        if self.pos > 0 && (self.writing || force) {
            send_recv_res!(
                &self.sgate,
                RecvGate::def(),
                GenFileOp::COMMIT,
                self.file_id(),
                self.pos
            )?;

            self.goff += self.pos;
            self.pos = 0;
            self.len = 0;
            self.writing = false;
        }
        Ok(())
    }

    fn delegate_ep(&mut self, ep_sel: Selector) -> Result<(), Error> {
        if ep_sel != self.delegated_ep {
            self.submit(true)?;
            let crd = CapRngDesc::new(CapType::OBJECT, ep_sel, 1);
            self.sess
                .delegate(crd, |s| s.push_word(GenFileOp::SET_DEST.val), |_| Ok(()))?;
            self.delegated_ep = ep_sel;
        }
        Ok(())
    }

    fn delegate_own_ep(&mut self) -> Result<(), Error> {
        self.mgate.activate()?;
        let ep_sel = self.mgate.ep().unwrap().sel();
        self.delegate_ep(ep_sel)
    }

    fn next_in(&mut self, len: usize) -> Result<usize, Error> {
        self.submit(false)?;
        if len == 0 {
            return Ok(0);
        }

        if self.pos == self.len {
            if !self.blocking && !self.receive_notify(FileEvent::INPUT, true)? {
                return Err(Error::new(Code::WouldBlock));
            }

            let mut reply = send_recv_res!(
                &self.sgate,
                RecvGate::def(),
                GenFileOp::NEXT_IN,
                self.file_id()
            )?;
            self.goff += self.len;
            self.off = reply.pop()?;
            self.len = reply.pop()?;
            self.pos = 0;
        }

        Ok(cmp::min(len, self.len - self.pos))
    }

    fn next_out(&mut self, len: usize) -> Result<usize, Error> {
        if len == 0 {
            return Ok(0);
        }

        if self.pos == self.len {
            if !self.blocking && !self.receive_notify(FileEvent::OUTPUT, true)? {
                return Err(Error::new(Code::WouldBlock));
            }

            let mut reply = send_recv_res!(
                &self.sgate,
                RecvGate::def(),
                GenFileOp::NEXT_OUT,
                self.file_id()
            )?;
            self.goff += self.len;
            self.off = reply.pop()?;
            self.len = reply.pop()?;
            self.pos = 0;
        }

        Ok(cmp::min(len, self.len - self.pos))
    }

    #[inline(never)]
    fn enable_notifications(&mut self) -> Result<(), Error> {
        if self.nb_state.is_some() {
            return Ok(());
        }

        let mut notify_rgate = Box::new(RecvGate::new(
            math::next_log2(NOTIFY_MSG_SIZE),
            math::next_log2(NOTIFY_MSG_SIZE),
        )?);
        notify_rgate.activate()?;
        let _notify_sgate = Box::new(SendGate::new(&notify_rgate)?);

        let crd = CapRngDesc::new(CapType::OBJECT, _notify_sgate.sel(), 1);
        self.sess.delegate(
            crd,
            |s| s.push_word(GenFileOp::ENABLE_NOTIFY.val),
            |_| Ok(()),
        )?;

        self.nb_state = Some(NonBlocking {
            notify_rgate,
            _notify_sgate,
            notify_received: FileEvent::empty(),
            notify_requested: FileEvent::empty(),
        });

        Ok(())
    }

    fn request_notification(&mut self, events: FileEvent) -> Result<(), Error> {
        let fid = self.file_id();
        let nb = self.nb_state.as_mut().unwrap();
        if !nb.notify_requested.contains(events) {
            send_recv_res!(
                &self.sgate,
                RecvGate::def(),
                GenFileOp::REQ_NOTIFY,
                fid,
                events.bits()
            )?;
            nb.notify_requested |= events;
        }
        Ok(())
    }

    #[inline(never)]
    fn receive_notify(&mut self, event: FileEvent, fetch: bool) -> Result<bool, Error> {
        // if we did not request a notification for this event yet, do that now
        if !self
            .nb_state
            .as_ref()
            .unwrap()
            .notify_requested
            .contains(event)
        {
            self.request_notification(event)?;
        }

        // if we did not receive the given event, check if there is a message
        let nb = self.nb_state.as_mut().unwrap();
        if !nb.notify_received.contains(event) {
            if let Some(msg) = nb.notify_rgate.fetch() {
                let mut imsg = GateIStream::new(msg, &nb.notify_rgate);
                let events = FileEvent::from_bits_truncate(imsg.pop::<u64>()?);
                nb.notify_received |= events;
                nb.notify_requested &= !events;
                // give credits back to sender
                imsg.reply_error(Code::None)?;
            }
        }

        // now check again if we have received this event; if not, we would block
        if !nb.notify_received.contains(event) {
            return Ok(false);
        }

        // okay, event received; remove it and continue
        if fetch {
            nb.notify_received &= !event;
        }
        Ok(true)
    }
}

impl Drop for GenericFile {
    fn drop(&mut self) {
        if !self.flags.contains(OpenFlags::NEW_SESS) {
            // we never want to invalidate the EP
            self.mgate.set_ep(None);
        }
    }
}

impl File for GenericFile {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn Any {
        self
    }

    fn fd(&self) -> Fd {
        self.fd
    }

    fn set_fd(&mut self, fd: Fd) {
        self.fd = fd;
    }

    fn session(&self) -> Option<Selector> {
        Some(self.sess.sel())
    }

    fn remove(&mut self) {
        // submit read/written data
        self.submit(false).ok();

        if !self.flags.contains(OpenFlags::NEW_SESS) {
            let file_id = self.id.unwrap();
            if let Some(fs) = Activity::own().mounts().get_by_id(self.fs_id.unwrap()) {
                fs.borrow_mut().close(file_id);
            }
        }
        else {
            // revoke EP cap
            if let Some(ep) = self.mgate.ep() {
                Activity::own()
                    .revoke(CapRngDesc::new(CapType::OBJECT, ep.sel(), 1), true)
                    .ok();
            }
        }

        // file sessions are not known to our resource manager; thus close them manually
        send_recv_res!(
            &self.sgate,
            RecvGate::def(),
            GenFileOp::CLOSE,
            self.file_id()
        )
        .ok();
    }

    fn stat(&self) -> Result<FileInfo, Error> {
        send_vmsg!(
            &self.sgate,
            RecvGate::def(),
            GenFileOp::STAT,
            self.file_id()
        )?;
        let reply = recv_reply(RecvGate::def(), Some(&self.sgate))?;
        let resp = reply.msg().get_data::<StatResponse>();
        FileInfo::from_response(resp)
    }

    fn path(&self) -> Result<String, Error> {
        let mut reply = send_recv_res!(
            &self.sgate,
            RecvGate::def(),
            GenFileOp::GET_PATH,
            self.file_id()
        )?;
        let path = reply.pop()?;

        let mounts = Activity::own().mounts();
        let mount_path = mounts
            .path_of_id(self.fs_id.unwrap())
            .ok_or(Error::new(Code::NotFound))?;
        Ok(mount_path.to_string() + "/" + path)
    }

    fn truncate(&mut self, length: usize) -> Result<(), Error> {
        self.submit(false)?;

        let mut reply = send_recv_res!(
            &self.sgate,
            RecvGate::def(),
            GenFileOp::TRUNCATE,
            self.file_id(),
            length
        )?;
        // reset position in case we were behind the truncated position
        self.goff = reply.pop()?;
        // we've lost access to the previous extent
        self.pos = 0;
        self.len = 0;
        Ok(())
    }

    fn file_type(&self) -> u8 {
        b'F'
    }

    fn delegate(&self, act: &ChildActivity) -> Result<Selector, Error> {
        let crd = CapRngDesc::new(CapType::OBJECT, self.sess.sel(), 2);
        self.sess.obtain_for(
            act.sel(),
            crd,
            |s| s.push_word(GenFileOp::CLONE.val),
            |_| Ok(()),
        )?;
        Ok(self.sess.sel() + 2)
    }

    fn serialize(&self, s: &mut StateSerializer<'_>) {
        s.push_word(self.flags.bits() as u64);
        s.push_word(self.sess.sel());
        s.push_word(self.fs_id.unwrap_or(!0) as u64);
    }

    fn is_blocking(&self) -> bool {
        self.blocking
    }

    fn set_blocking(&mut self, blocking: bool) -> Result<(), Error> {
        if !blocking {
            self.enable_notifications()?;
        }
        self.blocking = blocking;
        Ok(())
    }

    fn fetch_signal(&mut self) -> Result<bool, Error> {
        self.enable_notifications()?;

        self.receive_notify(FileEvent::SIGNAL, true)
    }

    fn check_events(&mut self, events: FileEvent) -> bool {
        if self.blocking {
            true
        }
        else {
            self.receive_notify(events, false).unwrap()
        }
    }
}

impl Seek for GenericFile {
    fn seek(&mut self, mut off: usize, mut whence: SeekMode) -> Result<usize, Error> {
        self.submit(false)?;

        if whence == SeekMode::CUR {
            off += self.goff + self.pos;
            whence = SeekMode::SET;
        }

        if whence != SeekMode::END
            && self.pos < self.len
            && off > self.goff
            && off < self.goff + self.len
        {
            self.pos = off - self.goff;
            return Ok(off);
        }

        let mut reply = send_recv_res!(
            &self.sgate,
            RecvGate::def(),
            GenFileOp::SEEK,
            self.file_id(),
            off,
            whence
        )?;

        self.goff = reply.pop()?;
        let off: usize = reply.pop()?;
        self.pos = 0;
        self.len = 0;
        Ok(self.goff + off)
    }
}

impl Read for GenericFile {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
        self.delegate_own_ep()?;

        let amount = self.next_in(buf.len())?;
        if amount > 0 {
            self.mgate
                .read(&mut buf[0..amount], (self.off + self.pos) as goff)?;
            self.pos += amount;
        }
        self.writing = false;
        Ok(amount)
    }
}

impl Write for GenericFile {
    fn flush(&mut self) -> Result<(), Error> {
        self.submit(false)
    }

    fn sync(&mut self) -> Result<(), Error> {
        self.flush().and_then(|_| {
            send_recv_res!(
                &self.sgate,
                RecvGate::def(),
                GenFileOp::SYNC,
                self.file_id()
            )
            .map(|_| ())
        })
    }

    fn write(&mut self, buf: &[u8]) -> Result<usize, Error> {
        self.delegate_own_ep()?;

        let amount = self.next_out(buf.len())?;
        if amount > 0 {
            self.mgate
                .write(&buf[0..amount], (self.off + self.pos) as goff)?;
            self.pos += amount;
        }
        self.writing = true;
        Ok(amount)
    }
}

impl Map for GenericFile {
    fn map(
        &self,
        pager: &Pager,
        virt: goff,
        off: usize,
        len: usize,
        prot: Perm,
        flags: MapFlags,
    ) -> Result<(), Error> {
        // TODO maybe check here whether self is a pipe and return an error?
        pager
            .map_ds(virt, len, off, prot, flags, &self.sess)
            .map(|_| ())
    }
}

impl HashInput for GenericFile {
    fn hash_input(&mut self, sess: &HashSession, len: usize) -> Result<usize, Error> {
        self.delegate_ep(sess.ep().sel())?;

        let mut remaining = len;
        while remaining > 0 {
            let amount = self.next_in(remaining)?;
            if amount == 0 {
                break;
            }

            sess.input(self.off + self.pos, amount)?;
            self.pos += amount;
            remaining -= amount;
        }
        Ok(len - remaining)
    }
}

impl HashOutput for GenericFile {
    fn hash_output(&mut self, sess: &HashSession, len: usize) -> Result<usize, Error> {
        self.delegate_ep(sess.ep().sel())?;

        let mut remaining = len;
        while remaining > 0 {
            let amount = self.next_out(remaining)?;
            if amount == 0 {
                break;
            }

            sess.output(self.off + self.pos, amount)?;
            self.pos += amount;
            remaining -= amount;
        }
        self.writing = true;
        Ok(len - remaining)
    }
}

impl fmt::Debug for GenericFile {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "GenFile[flags={:?}, sess={}, goff={:#x}, off={:#x}, pos={:#x}, len={:#x}]",
            self.flags,
            self.sess.sel(),
            self.goff,
            self.off,
            self.pos,
            self.len
        )
    }
}
